src/lib/comms/sol-lwm2m-bs-server.c

```diff
@@ -212,67 +213,95 @@ static const struct sol_coap_resource bootstrap_request_interface = {
-    switch (sec_mode) {
-    case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
-        SOL_NULL_CHECK(known_psks, NULL);
-        sec_mode_str = "Pre-Shared Key";
-        break;
-    case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
-        sec_mode_str = "Raw Public Key";
-        SOL_WRN("Raw Public Key security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
-        sec_mode_str = "Certificate";
-        SOL_WRN("Certificate security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
-        SOL_WRN("Bootstrap Server MUST use DTLS.");
-        return NULL;
-    default:
-        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
-        return NULL;
+
+    va_start(ap, num_sec_modes);
+
+    cipher_suites = calloc(num_sec_modes, sizeof(enum sol_socket_dtls_cipher));
+    SOL_NULL_CHECK_GOTO(cipher_suites, err_va_list);
+
+    sec_modes = calloc(num_sec_modes, sizeof(enum sol_lwm2m_security_mode));
+    SOL_NULL_CHECK_GOTO(sec_modes, err_cipher_suites);
+
+    for (i = 0; i < num_sec_modes; i++) {
+        sec_modes[i] = va_arg(ap, enum sol_lwm2m_security_mode);
+
+        switch (sec_modes[i]) {
+        case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
+            known_psks = va_arg(ap, struct sol_lwm2m_security_psk **);
+            SOL_NULL_CHECK_GOTO(known_psks, err_sec_modes);
+
+            cipher_suites[i] = SOL_SOCKET_DTLS_CIPHER_PSK_AES128_CCM8;
+            break;
+        case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
+            my_rpk = va_arg(ap, struct sol_lwm2m_security_rpk *);
+            SOL_NULL_CHECK_GOTO(my_rpk, err_sec_modes);
+            known_pub_keys = va_arg(ap, struct sol_blob **);
+            SOL_NULL_CHECK_GOTO(known_pub_keys, err_sec_modes);
+
+            cipher_suites[i] = SOL_SOCKET_DTLS_CIPHER_ECDHE_ECDSA_AES128_CCM8;
+            break;
+        case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
+            SOL_WRN("Certificate security mode is not supported yet.");
+            goto err_sec_modes;
+        case SOL_LWM2M_SECURITY_MODE_NO_SEC:
+            SOL_WRN("Bootstrap Server MUST use DTLS.");
+            goto err_sec_modes;
+        default:
+            SOL_WRN("Unknown DTLS Security Mode: %d", sec_modes[i]);
+            goto err_sec_modes;
+        }
     }
 
     server = calloc(1, sizeof(struct sol_lwm2m_bootstrap_server));
-    SOL_NULL_CHECK(server, NULL);
+    SOL_NULL_CHECK_GOTO(server, err_sec_modes);
 
     //LWM2M Bootstrap Server MUST always use DTLS
-    if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
-        sol_vector_init(&server->known_psks, sizeof(sol_lwm2m_security_psk));
-
-        SOL_VECTOR_FOREACH_IDX (known_psks, given_psk, i) {
-            psk = sol_vector_append(&server->known_psks);
-            SOL_NULL_CHECK_GOTO(psk, err_psk);
-            psk->id = sol_blob_ref(given_psk->id);
-            psk->key = sol_blob_ref(given_psk->key);
+    for (i = 0; i < num_sec_modes; i++) {
```

Comment:
I don't get it. I don't have the `server` allocated in the `switch` there, and I don't have the `sol_coap_server server->coap` allocated in the `for` there. I need both those objects set to use `_security_add()`.